Updates to PhysX Collider, PhysX Shape Collider, and PhysX Rigid Body… (#1359)
o3de/o3de.org: content/docs/user-guide/components/reference/physx/collider.md

{{< /note >}}

## Provider

[PhysX Gem](/docs/user-guide/gems/reference/physics/nvidia/physx/)

## Properties 

![PhysX Collider component interface.](/images/user-guide/components/reference/physx/physx-collider-ui-01.png)

### Base properties

| Property | Description | Value | Default |
| - | - | - | - |
| **Collision Layer** | Assigns the collider to a collision layer. Collision layers can be used to restrict physical interactions between PhysX objects. | Any collision layer defined the project [Collision Layers](/docs/user-guide/interactivity/physics/nvidia-physx/configuring/configuration-collision-layers/).  | `Default` |
| **Collides With** | Assigns the collider to a collision group. Collision groups contain the collision layers that this collider can collide with. | Any collision group defined the project [Collision Groups](/docs/user-guide/interactivity/physics/nvidia-physx/configuring/configuration-collision-groups/). | `All` |

> | **Trigger** | If enabled, this collider functions as a trigger. A trigger performs a quick overlap test and does not apply forces or return contact point information. Use this to speed-up PhysX computations where a simple overlap test between colliders is sufficient. Triangle meshes are not supported as triggers. | Boolean | `Disabled` |

Reviewer: ```suggestion
| **Trigger** | If enabled, this collider functions as a trigger. A trigger checks if other colliders are overlapping with this collider. It does not apply forces or return contact point information. Use this to speed-up PhysX computations where a simple overlap test between colliders is sufficient. Triangle meshes are not supported as triggers. | Boolean | `Disabled` |
```

- I suggest expanding on what "overlap test" means. 
- If you accept the point above, consider splitting up the sentence.

Author: I'm going to make a small change to the suggestion. I think splitting the sentence is good, but the most critical bit of information is that overlap test are very quick.

```suggestion
| **Trigger** | If enabled, this collider functions as a trigger. A trigger performs quick overlap tests with other colliders. Triggrers don't apply forces or return contact point information. Use this to speed up PhysX computations in which a simple overlap test between colliders is sufficient. Triangle meshes are not supported as triggers. | Boolean | `Disabled` |
```

| **Simulated** | If enabled, this collider is included in the physics simulation. | Boolean | `Enabled` |
| **In Scene Queries** | If enabled, this collider can be queried for raycasts, shapecasts, and overlap. | Boolean | `Enabled` |

> | **Offset** | Sets a local offset position for the collider relative to the entity. | Vector3: -Infinity to Infinity | X: `0.0`, Y: `0.0`, Z: `0.0` |

Reviewer: ```suggestion
| **Offset** | Sets the collider's local offset position relative to the entity. | Vector3: -Infinity to Infinity | X: `0.0`, Y: `0.0`, Z: `0.0` |
```
- Clarify that "relative to the entity" applies to the local offset position.

> | **Rotation** | Sets a local rotation for the collider around the **Offset** of the PhysX collider component. | Vector3: -180 to 180 | X: `0.0`, Y: `0.0`, Z: `0.0` |

Reviewer: ```suggestion
| **Rotation** | Sets a local rotation for the collider around the **Offset** of the PhysX Collider component. | Vector3: -180 to 180 | X: `0.0`, Y: `0.0`, Z: `0.0` |
```

Author: component name.

> | **Rotation** | Sets a local rotation for the collider around the **Offset** of the PhysX collider component. | Vector3: -180 to 180 | X: `0.0`, Y: `0.0`, Z: `0.0` |

Author: Funny, I actually went through and removed all the decimals after initially adding them. I'll take these just to be consistent.

> | **Collision Layer** | Assigns the collider to a collision layer. Collision layers can be used to restrict physical interactions between PhysX objects. | Any collision layer defined in the project's [collision layers](/docs/user-guide/interactivity/physics/nvidia-physx/configuring/configuration-collision-layers/).  | `Default` |
> | **Collides With** | Assigns the collider to a collision group. Collision groups contain the collision layers that this collider can collide with. | Any collision group defined in the project's [collision groups](/docs/user-guide/interactivity/physics/nvidia-physx/configuring/configuration-collision-groups/). | `All` |

Author: Changing these to match the shape collider page, and because I think they are the GUI labels and will ultimately be the topic titles.

```suggestion
| **Collision Layer** | Assigns the collider to a collision layer. Collision layers can be used to restrict physical interactions between PhysX objects. | Any collision layer defined in the project's [Collision Layers](/docs/user-guide/interactivity/physics/nvidia-physx/configuring/configuration-collision-layers/).  | `Default` |
| **Collides With** | Assigns the collider to a collision group. Collision groups contain the collision layers that this collider can collide with. | Any collision group defined in the project's [Collision Groups](/docs/user-guide/interactivity/physics/nvidia-physx/configuring/configuration-collision-groups/). | `All` |
```

| **Library (Physics Materials)** | The physics material library of the project. | A `.physmaterial` library product asset. | The global project `.physmaterial` library. |
| **Slots (Physics Materials)** | Choose a physics material for each material of this collider. Physics materials define physical properties for the surface such as dynamic and static friction, and density. A collider can have multiple physics materials assigned. | Physics material(s) from the assigned `.physmaterial` library | `<Default Physics Material>` |
| **Tag** | Sets a tag for this collider. Tags can be used to quickly identify components in script or code. | String |  |

> | **Rest offset** | Sets the minimum distance between this collider and other colliders when this collider is at rest. PhysX bodies come to rest separated by the sum of their **Rest offset** values. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50 | `0.0` |

Reviewer: Building off of FiniteStateGit's suggestion.
I suggest adding "position" to specify what's being separated. Alternatively, you can say: "PhysX bodies come to rest at a position separated by..."
```suggestion
| **Rest offset** | Sets the minimum distance between this collider and other colliders when this collider is at rest. PhysX bodies's rest positions are separated by the sum of their **Rest offset** values. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50.0 | `0.0` |
```

> | **Rest offset** | Sets the minimum distance between this collider and other colliders when this collider is at rest. PhysX bodies come to rest separated by the sum of their **Rest offset** values. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50 | `0.0` |

Author: I don't think the term rest position should be used here because it implies a specific default/start position for an animation or simulation. This is the body coming to rest in some undefined position due to friction/damping.

> | **Rest offset** | Sets the minimum distance between this collider and other colliders when this collider is at rest. PhysX bodies come to rest separated by the sum of their **Rest offset** values. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50 | `0.0` |

Reviewer: How about: "PhysX bodies come to rest at a position separated by the sum of their **Rest offset** values."

I think it helps to clarify _what_ is "separated by the sum of their **Rest offset** values"

> | **Rest offset** | Sets the minimum distance between this collider and other colliders when this collider is at rest. PhysX bodies come to rest separated by the sum of their **Rest offset** values. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50 | `0.0` |

Author: Thinking about this, this is actually a poorly named property. Technically, the concept of rest really only applies to dynamic colliders, but every collider type uses this offset. Maybe this is better. I really want to avoid using the terms 'rest' and 'position' in close proximity in this context because of what it implies.

```suggestion
| **Rest offset** | Sets the minimum distance between this collider and other colliders. Although this property applies to all colliders, it is particularly important for dynamic colliders. Dynamic colliders are at rest when the forces affecting them drop below the **Sleep threshold** of their rigid body component. When a dynamic collider comes to rest while in contact with any other collider, the colliders are separated by the sum of their **Rest offset** values. **Rest offset** values that are too large might make dynamic entities appear to float. Negative **Rest offset** values might make dynamic entities appear to intersect other entities on collision. You might need to adjust this value in scenarios where the collider does not closely match the render mesh of the entity to improve the visual fidelity of PhysX simulations. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50.0 | `0.0` |
```

> | **Rest offset** | Sets the minimum distance between this collider and other colliders when this collider is at rest. PhysX bodies come to rest separated by the sum of their **Rest offset** values. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50.0 | `0.0` |

Author: ```suggestion
| **Rest offset** | Sets the minimum distance between this collider and other colliders. Although this property applies to all colliders, it is particularly important for dynamic colliders. Dynamic colliders are at rest when the forces affecting them drop below the **Sleep threshold** of their rigid body component. When a dynamic collider comes to rest while in contact with any other collider, the colliders are separated by the sum of their **Rest offset** values. **Rest offset** values that are too large might make dynamic entities appear to float. Negative **Rest offset** values might make dynamic entities appear to intersect. You might need to adjust this value in scenarios where the collider does not closely match the render mesh of the entity. The **Rest offset** value must be less than the **Contact offset** value. | Float: -Infinity to 50.0 | `0.0` |
```

| **Draw Collider** | If enabled, the collider is displayed in the viewport. | Boolean | `Enabled` |
| **Edit** | Enter collider component edit mode to adjust properties of the collider with manipulators in the viewport. |  |  |

### PhysicsAsset shape properties

![PhysX Collider component interface, Physics Asset.](/images/user-guide/components/reference/physx/physx-collider-ui-02.png)

| Property | Description | Value | Default |
| - | - | - | - |
| **PhysX Mesh** | Assigns a `.pxmesh` collider product asset for this collider. For more information on creating PhysX mesh asset colliders, refer to [Process PhysX Collider Assets](/docs/learning-guide/tutorials/assets/physx-colliders/). | Product asset `.pxmesh` PhysX mesh. |  |

> | **Asset Scale** | Scale the collider shape independent of the entity. | Vector3: 0 to Infinity | X: `1.0`, Y: `1.0`, Z: `1.0` |

Reviewer: Builds off of FiniteStateGit's suggestion.
```suggestion
| **Asset Scale** | Scales the collider shape independent of the entity. | Vector3: 0.0 to Infinity | X: `1.0`, Y: `1.0`, Z: `1.0` |
```

### Sphere shape properties

![PhysX Collider component interface, Sphere.](/images/user-guide/components/reference/physx/physx-collider-ui-03.png)

| Property | Description | Value | Default |
| - | - | - | - |
| **Radius** | Radius multiplier of the sphere collider. The size of the sphere primitive is the **Radius** value multiplied by the largest value in the **Scale** property in the entity's [Transform](/docs/user-guide/components/reference/transform/) component. | Float: 0 to Infinity | `0.5` |

### Box shape properties

![PhysX Collider component interface, Box.](/images/user-guide/components/reference/physx/physx-collider-ui-04.png)

| Property | Description | Value | Default |
| - | - | - | - |
| **Dimensions** | Width, depth, and height of the box collider. | Vector3: 0 to Infinity | X: `1.0`, Y: `1.0`, Z: `1.0` |

### Capsule shape properties

![PhysX Collider component interface, Box.](/images/user-guide/components/reference/physx/physx-collider-ui-05.png)

| Property | Description | Value | Default |
| - | - | - | - |

> | **Height** | Height of the capsule primitive shape. The **Height** value of the capsule must be at least twice the **Radius** value. For example, if the **Radius** of the capsule is `5.0`, the minimum **Height** is `10.0`. | Float: 0 to Infinity | `1.0` |

Reviewer: ```suggestion
| **Height** | Height of the capsule collider. The **Height** value of the capsule must be at least twice the **Radius** value. For example, if the **Radius** of the capsule is `5.0`, the minimum **Height** is `10.0`. | Float: 0.0 to Infinity | `1.0` |
```

For consistency. You used "sphere collider" and "box collider" in the earlier property descriptions.

> | **Radius** | Radius of the capsule primitive shape. The **Radius** value of the capsule must be no greater than half the **Height** value. For example, if the **Height** of the capsule is `10.0`, the maximum **Radius** is `5.0`. | Float: 0 to Infinity | `0.25` |

Author: Based on Chanmosq'a suggestion above.

```suggestion
| **Radius** | Radius of the capsule collider. The **Radius** value of the capsule must be no greater than half the **Height** value. For example, if the **Height** of the capsule is `10.0`, the maximum **Radius** is `5.0`. | Float: 0.0 to Infinity | `0.25` |
```

![PhysX Collider component mode capsule resize manipulator](/images/user-guide/components/reference/physx/physx-collider-resize-capsule.png)

### Resize (Physics Asset Shape)

Physics Asset resize mode has a three axis scale manipulator.

![PhysX Collider component mode physics asset resize manipulator](/images/user-guide/components/reference/physx/physx-collider-resize-physicsasset.png)

### Offset

Offset mode has a three axis translate manipulator.

![PhysX Collider component mode offset manipulator](/images/user-guide/components/reference/physx/physx-collider-offset-mode.png)

### Rotation

Rotation mode has a three axis rotate manipulator.

![PhysX Collider component mode rotate manipulator](/images/user-guide/components/reference/physx/physx-collider-rotate-mode.png)

### Collider component mode hotkeys

The following navigation hotkeys are available in collider component mode.

| Hotkey | Action |
| - | - |
| **1** | Resize mode. |
| **2** | Offset mode. |
| **3** | Rotation mode. |
| **CTRL + Mouse Wheel Up** | Next mode. |
| **CTRL + Mouse Wheel Down** | Previous mode. |

> | **R** | Reset current mode. This is effectively an undo operation. You can step through the Resize, Offset, and Rotation modes and press R to reset changes to the current mode. |

Reviewer: ```suggestion
| **R** | Reset current mode. This is effectively an undo operation. You can step through the Resize, Offset, and Rotation modes and press **R** to reset changes to the current mode. |
```

| **ESC** | Exit component mode. |

## Colliders as triggers

> Triggers allow colliders to perform efficient overlap tests. Colliders marked as triggers won't have forces applied when they intersect with another collider. This is useful for detecting when something enters a certain area or when two objects overlap. Use Lua or Script Canvas to detect overlap.

Reviewer: ```suggestion
Triggers allow colliders to perform efficient overlap tests. Colliders marked as triggers won't be affected by forces when they intersect with another collider. This is useful for detecting when something enters a certain area or when two objects overlap. Use Lua or **Script Canvas** to detect overlap.
```

Built off of FiniteStateGit's review.
